### PR TITLE
Support bfloat16 init

### PR DESCRIPTION
**Pull request by t-vi**
With this patch, I can init in bfloat16, this helps run 7B on my RTX consumer card.

### CODE REVIEW

**Comment by lantiga** [comment]
Great, I would have it in utils for now, and then we upstream to PL/Fabric

**Comment by t-vi** [comment]
OK, so I'll add a bfloat16 command line option for generate tomorrow and update the readme...

**Comment by t-vi** [comment]
If the CI doesn't complain, I think I'm reasonably happy with it.

**Review by lantiga — APPROVED** [review]
Looks great, let's hear from @awaelchli too

File: lit_llama/utils.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+            with EmptyInitOnDevice("cuda", dtype=torch.bfloat16):
+               model = LLaMA.from_name('7B')
+            model.load_state_dict(torch.load('llama-lit/7B/state_dict.pth'))"""
+
+        self.quantization_mode = quantization_mode
+        if self.quantization_mode == 'llm.int8':
+            if device.type != "cuda":
+                raise ValueError("Quantization is only supported on the GPU.")
+            from .quantization import Linear8bitLt
+            self.Linear8bitLt = Linear8bitLt
```
Comment (awaelchli) [inline lit_llama/utils.py:25]: shouldn't this be moved to `__enter__` directly?

File: lit_llama/utils.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
+import torch.utils._device
+
+
+class EmptyInitOnDevice(torch.overrides.TorchFunctionMode):
+    def __init__(self, device=None, dtype=None, quantization_mode=None):
+        """
+        Create tensors with given device and dtype and don't run initialization
+           (but instead use "empty tensors", i.e. uninitialized memory).
```
Comment (awaelchli) [inline lit_llama/utils.py:8]: I don't see where the creation of empty tensors happens.

Comment (t-vi) [reply lit_llama/utils.py:32]: PyTorch uses torch.empty to create the buffers for modules and then uses the overriden `toch.nn.init.*` to initialize:

https://github.com/pytorch/pytorch/blob/d47a4bf53fcde103ea6ea52eef66d999f707f61f/torch/nn/modules/linear.py#L96-L98

File: lit_llama/utils.py (hunk 1 above)
Comment (t-vi) [reply lit_llama/utils.py:49]: I have reservations importing in `__enter__` but maybe that is just my excellent taste in code style and cannot be generalized.

**Review by awaelchli — APPROVED** [review]

**Pull request merged**